browser/go/providerwrapper/providerwrapper.go

```diff
@@ -0,0 +1,183 @@
+	//         jsonrpc: string;
+	//     }
+	//
+	//     interface JSONRPCResponseError {
+	//         message: string;
+	//         code: number;
+	//     }
+	//
+	//     interface JSONRPCResponsePayload {
+	//         result: any;
+	//         id: number;
+	//         jsonrpc: string;
+	//         error?: JSONRPCResponseError;
+	//     }
+	//
+	//     type JSONRPCErrorCallback = (err: Error | null, result?: JSONRPCResponsePayload) => void;
+	//
+	//     sendAsync(payload: JSONRPCRequestPayload, callback: JSONRPCErrorCallback): void;
+	//
+
+	// Set up payload
+	payload := map[string]interface{}{
+		"jsonrpc": "2.0",
+		"id":      rand.Intn(math.MaxInt32),
+		"method":  method,
+	}
+	if len(args) > 0 {
+		// Convert args to a value that is compatible with syscall/js. Since we don't
+		// know the underlying type of args, the only reliable way to do this is to
+		// convert to and from JSON.
+		convertedParams, err := jsutil.InefficientlyConvertToJS(args)
+		if err != nil {
+			return fmt.Errorf("invalid args for JSON payload: %s", err.Error())
+		}
+		payload["params"] = convertedParams
+	}
+
+	// Set up the callback function
+	resultChan := make(chan js.Value, 1)
+	errChan := make(chan error, 1)
+	var callback js.Func
+	callback = js.FuncOf(func(this js.Value, args []js.Value) interface{} {
+		defer callback.Release()
+		go func() {
+			if len(args) == 0 {
+				errChan <- errors.New("JSONRPCErrorCallback called with no arguments")
+				return
+			}
+			jsError := args[0]
+			if !jsutil.IsNullOrUndefined(jsError) {
+				errChan <- js.Error{
+					Value: jsError,
+				}
+				return
+			}
+			if len(args) < 2 {
+				errChan <- errors.New("JSONRPCErrorCallback called with null/undefined error but no results")
+				return
+			}
+			resultChan <- args[1]
+			return
+		}()
+		return nil
+	})
+
+	// Call sendAsync and use select to wait for the results.
+	c.provider.Call("sendAsync", payload, callback)
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	case err := <-errChan:
+		return err
+	case jsResult := <-resultChan:
+		if rpcErr := jsResult.Get("error"); !jsutil.IsNullOrUndefined(rpcErr) {
+			return jsErrorToRPCError(rpcErr)
+		}
+		if result == nil {
+			return nil
+		}
+		if err := jsutil.InefficientlyConvertFromJS(jsResult.Get("result"), result); err != nil {
+			return fmt.Errorf("could not decode JSON RPC response: %s", err.Error())
+		}
+		return nil
+	}
+}
+
+// BatchCall sends all given requests as a single batch and waits for the server
+// to return a response for all of them. The wait duration is bounded by the
+// context's deadline.
+//
+// In contrast to CallContext, BatchCallContext only returns errors that have occurred
+// while sending the request. Any error specific to a request is reported through the
+// Error field of the corresponding BatchElem.
+//
+// Note that batch calls may not be executed atomically on the server side.
+func (c *RPCClient) BatchCallContext(ctx context.Context, b []rpc.BatchElem) error {
```

Comment:
I didn't implement `BatchCallContext` or `EthSubscribe` since they don't seem to actually be used in the browser. We can implement them later if needed.

Reply:
Sounds good. We might use `EthSubscribe` is we decide to use a WS subscription for watching new blocks, but we don't use it now.